flink-core/src/main/java/org/apache/flink/core/state/StateFutureImpl.java

```diff
@@ -61,23 +66,34 @@ public <U> StateFuture<U> thenApply(Function<? super T, ? extends U> fn) {
                 return StateFutureUtils.completedFuture(r);
             } else {
                 StateFutureImpl<U> ret = makeNewStateFuture();
-                completableFuture.thenAccept(
-                        (t) -> {
-                            callbackRunner.submit(
-                                    () -> {
-                                        ret.complete(fn.apply(t));
-                                        callbackFinished();
-                                    });
-                        });
+                completableFuture
+                        .thenAccept(
+                                (t) -> {
+                                    callbackRunner.submit(
+                                            () -> {
+                                                ret.complete(fn.apply(t));
+                                                callbackFinished();
+                                            });
+                                })
+                        .exceptionally(
```

Comment:
Shall we do the `exceptionally` here? IIUC, we won't call `completeExceptionally` right?
I was thinking the callback should be protected by:
```
completableFuture
                        .thenAccept(
                                (t) -> {
                                    try {
                                        callbackRunner.submit(
                                                () -> {
                                                    ret.complete(fn.apply(t));
                                                    callbackFinished();
                                                });
                                      } catch (Throwable e) { exceptionHandler.handleException }
                                })
```
Right?

Reply:
`exceptionally` and `try/catch` are equivalent, this is an explanation regarding `exceptionally`. I prefer writing it in terms of `exceptionally` to reduce nested levels.

> Returns a new CompletableFuture that is completed when this CompletableFuture completes, with the result of the given function of the exception triggering this CompletableFuture's completion when it completes exceptionally; otherwise, if this CompletableFuture completes normally, then the returned CompletableFuture also completes normally with the same value

Reply:
thanks for the explanation, I'm fine with current implementation